support-modules/acquisition-events/src/main/scala/com/gu/support/acquisitions/ga/GoogleAnalyticsServiceImpl.scala

```diff
@@ -0,0 +1,255 @@
+    val tid = transactionId.getOrElse(UUID.randomUUID().toString)
+    val goExp = buildOptimizeTestsPayload(acquisition.abTests)
+
+    // clientId cannot be empty or the call will fail
+    sanitiseClientId(gaData.clientId).map {
+      clientId =>
+        val productName = getProductName(acquisition)
+        val conversionCategory = getConversionCategory(acquisition)
+        val productCheckout = getProductCheckout(acquisition)
+        val body = Map(
+          "v" -> "1", //Version
+          "cid" -> clientId,
+          "tid" -> gaPropertyId,
+          "dh" -> gaData.hostname,
+          "uip" -> gaData.clientIpAddress.getOrElse(""), // IP Override
+          "ua" -> gaData.clientUserAgent.getOrElse(""), // User Agent Override
+
+          // Custom Dimensions
+          "cd12" -> acquisition.campaignCode.getOrElse(""), // Campaign code
+          "cd16" -> buildABTestPayload(acquisition.abTests), //'Experience' custom dimension
+          "cd17" -> acquisition.paymentProvider.getOrElse(""), // Payment method
+          "cd19" -> acquisition.promoCode.getOrElse(""), // Promo code
+          "cd25" -> acquisition.labels.exists(_.contains("REUSED_EXISTING_PAYMENT_METHOD")), // usedExistingPaymentMethod
+          "cd26" -> acquisition.labels.exists(_.contains("gift-subscription")), // gift subscription
+          "cd27" -> productCheckout,
+          "cd30" -> acquisition.labels.exists(_.contains("corporate-subscription")), // corporate subscription
+
+          // Custom metrics
+          "cm10" -> getSuccessfulSubscriptionSignUpMetric(conversionCategory),
+
+          // Google Optimize Experiment Id
+          "xid" -> goExp.testNames,
+          "xvar" -> goExp.variantNames,
+
+          // The GA conversion event
+          "t" -> "event",
+          "ec" -> conversionCategory.name, //Event Category
+          "ea" -> productName, //Event Action
+          "el" -> camelCase(acquisition.paymentFrequency.value), //Event Label
+          "ev" -> acquisition.amount.map(_.toInt).getOrElse(0).toString, //Event Value is an Integer
+
+          // Enhanced Ecommerce tracking https://developers.google.com/analytics/devguides/collection/protocol/v1/devguide#enhancedecom
+          "ti" -> tid,
+          "tcc" -> acquisition.promoCode.getOrElse(""), // Transaction coupon.
+          "pa" -> "purchase", //This is a purchase
+          "pr1nm" -> productName, // Product Name
+          "pr1ca" -> conversionCategory.description, // Product category
+          "pr1pr" -> annualisedValue, // Product Price - we are tracking the annualised value here as this is what goes into the revenue metric
+          "pr1qt" -> "1", // Product Quantity
+          "pr1cc" -> acquisition.promoCode.getOrElse(""), // Product coupon code.
+          "pr1cm15" -> acquisition.amount.toString, // Custom metric 15 - purchasePrice
+          "cu" -> acquisition.currency.toString // Currency
+        )
+
+        body
```

Comment:
body is a Map[String, Any] which means we are comparing an Any with `""` and calling toString on something of unknown type, which is probably not what we want.

Reply:
this is the case with the original code, there are a few bools and an option. I don't think it would hurt to stringify these to make it a `Map[String,String]

Reply:
maybe not, and if it matches the existing functionality we can keep it.  personally I'd prefer to have the toString explicitly on the relevant lines, and also not to toString scala types to put into external systems in case they are overridden or refactored to use another type later.
Regarding the option, actually I was thinking about whether we can get rid of it completely, but if we are relying on the downstream being `"Some(blah)"` then that may not be possible

Reply:
I've pushed up a change to stringify those values. I've defaulted the option to `""`, which is what we do in various other places

Reply:
perfect (as long as it's not going to break existing consumers of the GA data)